### PR TITLE
Add arm-linux-gnueabi-gcc build

### PR DESCRIPTION
**Pull request by mpg**
## Description

Add a build with arm-linux-gnueabi-gcc - compared to existing Arm builds, this allows building the programs and tests, in order to catch platform-dependent issues there too (previously we only cross-built the library to Arm).

See https://github.com/ARMmbed/mbedtls/pull/3449#issuecomment-675313720

See also. https://github.com/ARMmbed/mbedtls/issues/3299

## Status
**READY**

## Requires Backporting
Yes - testing improvement 

- [x] 2.16 #4855
- [x] 2.2x #4856

## Additional comments

* [x] The newly added test on Travis is expected to fail until #3449 is merged.
* [x] This PR depends on https://github.com/ARMmbed/mbedtls-test/pull/190 or https://github.com/ARMmbed/mbedtls-test/pull/191 getting merged and the new Docker image deployed.

### CODE REVIEW

**Comment by hanno-arm** [comment]
Thanks @mpg for this PR! Can we go one step further and use QEMU and syscall emulation to actually run our test suites? I've done this locally lately and it didn't require any change to the test scripts - once the tooling is setup correctly, the system will detect the target architecture for binaries automatically and transparently hook in QEMU.

**Comment by mpg** [comment]
Yes, sure, that would be a great improvement! But I'd rather take an incremental approach and add qemu in a follow-up PR - feel free to submit one if you already have something working locally.

(Also, I'm not sure if you're aware, but at the moment we can't easily update the Docker images used by our Jenkins CI, there are a couple of things we need to address before we can do that again, see https://github.com/ARMmbed/mbedtls-test/issues/150 - so for now we can only install additional software on Travis.)

**Comment by hanno-arm** [comment]
@mpg Ok, let's leave it to another PR.

> (Also, I'm not sure if you're aware, but at the moment we can't easily update the Docker images used by our Jenkins CI, there are a couple of things we need to address before we can do that again, see ARMmbed/mbedtls-test#150 - so for now we can only install additional software on Travis.)

Yes, @dgreen-arm told me about it, and that was the reason I didn't pursue adding Arm tests to the CI anymore (my main interest is getting tests for #1424), but I overlooked Travis, which is also a good start.

**Comment by mpg** [comment]
The CI is failing because Jenkins has a check to make sure it can run all the components, which makes sense, but prevents us from merging this PR before we've updated the Docker images, so this PR also depends on https://github.com/ARMmbed/mbedtls-test/issues/150

File: tests/scripts/all.sh
Comment (ronald-cron-arm) [inline tests/scripts/all.sh:1814]:
```suggestion
    # component_build_arm_linux_gnueabi_gcc_arm5vte
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks overall good to me. I have two minor comments and this needs to be rebased.

**Comment by mpg** [comment]
@ronald-cron-arm Thanks for your review! I've rebased to resolve the conflicts, and applied the suggested fix in the original commit since I was rebasing anyway.

**Comment by ronald-cron-arm** [comment]
From the CI it seems that the check to run `build_arm_linux_gnueabi_gcc_arm5vte` only if the `gcc-arm-llinux-gnueabi` is installed does not work.

**Comment by mpg** [comment]
> From the CI it seems that the check to run `build_arm_linux_gnueabi_gcc_arm5vte` only if the `gcc-arm-llinux-gnueabi` is installed does not work.

Jenkins correctly reports "Unable to run all.sh components: build_arm_linux_gnueabi_gcc_arm5vte", so I think the check works and the Jenkins job notices it can't run this component (as opposed to trying to run it and then failing) but it was also designed to not silently skip components it can't run and fail instead. I think that part is fully expected as long as we haven't installed the toolchain in our Docker image (which is why this PR was stalled).

I'm more worried about the failure in Travis, apparently our install is not complete. (The test passes on my machine, so I must have the right package but now I need to identify which one it is.)

**Comment by mpg** [comment]
OK, Travis is passing now. Now we're waiting for https://github.com/ARMmbed/mbedtls-test/issues/165 to update the Docker image.

**Comment by ronald-cron-arm** [comment]
> Jenkins correctly reports "Unable to run all.sh components: build_arm_linux_gnueabi_gcc_arm5vte", so I think the check works and the Jenkins job notices it can't run this component (as opposed to trying to run it and then failing) but it was also designed to not silently skip components it can't run and fail instead. I think that part is fully expected as long as we haven't installed the toolchain in our Docker image (which is why this PR was stalled).

ok thanks for the explanation.

**Review by ronald-cron-arm — DISMISSED** [review]
LGTM

**Review by gilles-peskine-arm — APPROVED** [review]

**Comment by mpg** [comment]
Do you mind if I rebase this on recent development? This would avoid the distracting failure of the API/ABI checker.

**Comment by gilles-peskine-arm** [comment]
Please go ahead and rebase.

**Review by gilles-peskine-arm — APPROVED** [review]
Re-approving after rebase.

Now that the necessary compiler is present on Jenkins, I have verified that the PR job ran it as expected.

**Review by mstarzyk-mobica — APPROVED** [review]

**Review by lukgni — APPROVED** [review]

**Pull request merged**
